### ISSUE TITLE
[Bug] vapoursynth - frame inaccuracies when selecting frame ranges / exact frame

### ISSUE THREAD

**Issue by adworacz** [vapoursynth/vapoursynth#635]
TL;DR - Incorrect frames are being returned when requesting a segment vs loading a video from start to end.

This is a little involved of a bug report, so bear with me as I add images and upload example videos

While working on Vapoursynth support for [Av1an](https://github.com/master-of-zen/Av1an/issues/173), I noticed some odd glitching upon playback. I thought it was Av1an for a while, but further investigation points to Vapoursynth as the cause. 

Script:
```
#!/usr/bin/env python
import vapoursynth as vs

core = vs.core
source = "./video.mkv"

def sample_extract(src, shots=18, shot_duration=5):
    """returns a sample clip of <shots> scenes of <shot_duration> seconds respectively"""
    return core.std.SelectEvery(src, src.num_frames//shots, range(0,round(src.fps*shot_duration))).std.AssumeFPS(src)

clip = core.ffms2.Source(source)
# clip = core.lsmas.LWLibavSource(source)
clip = clip.std.CropAbs(1920, 800, 0, 140)
clip = sample_extract(clip)
clip.set_output()
```

Now, when this script is rendered end to end / frames requested in order, everything is fine. But when using programs like `vspreview` or using the `--start --end` parameters of vspipe, things get wonky.

Here is an example of the exact same script loaded into two different versions of `vspreview`. The one on the right was manually set to frame 0, and then "fast forwarded" (held the button down) until the problematic frame (194, which is supposed to be a scene cut) was reached.

The image on the left shows the "out of order" request behavior, since vspreview used a bookmark, and the image on the right shows the "in order" (correct) behavior. Note how the frame number (bottom left corner) is exactly the same in both (194).
![2020-10-12_12:06:41_maim](https://user-images.githubusercontent.com/561689/95789906-3e4ea780-0c93-11eb-8b46-1e87678b11ed.png)

If I take the script and render it end to end, here's the surrounding range of frames in Wobbly
![2020-10-12_13:39:11_maim](https://user-images.githubusercontent.com/561689/95789994-6b9b5580-0c93-11eb-94c5-a24d7b938304.png)

But, if I render it using Av1an, which splits up a video into chunks based on scene detection (I added VS support, and its using `vspipe --start xxx --end yyy` to do this), I get the following in Wobbly:
![2020-10-12_13:45:49_maim](https://user-images.githubusercontent.com/561689/95790134-a7ceb600-0c93-11eb-9916-0fa6aedf2e38.png)

Lastly, if I just use vspipe + mpv, selecting frame 194 manually, you can see that I get the incorrect frame of the astronaut instead of the spaceship computer screen:
`vspipe -y -s 194 -e 194 vsScript.py - | mpv --pause --demuxer-rawvideo-codec=y4m -`

Result:
![mpv-shot0012](https://user-images.githubusercontent.com/561689/95790298-ef554200-0c93-11eb-85ad-b16126228979.png)

This happens with both FFMS2 and LSMASH (see above script commenting out one or the other, both exhibit the same issue). Additionally this happens with vspipe AND vspreview, so I think it something related to the core of Vapoursynth, not either of those programs.

If the source video is desired, let me know. It's going to take a while to upload, as a 27GB video-only stream from a Bluray, so I only want to take time to upload it if its worth it.

**Comment by adworacz** [comment]
Sorry, this is using Vapoursynth 52, on Manjaro Linux, kernel 5.8.11 if that's useful at all.

**Comment by theChaosCoder** [comment]
Check if ffms2/lsmash is frame accurate with your source file https://gist.github.com/dubhater/3a2c8a59841cae49ecae25cd47ff78d2

**Comment by adworacz** [comment]
Ahhhh interesting, I was operating under the assumption that those sources guaranteed frame accuracy, especially for AVC/H264 content, but maybe that’s a bad assumption. 

I’m running the test with ffms2 right now, and it’s spitting out a lot of ‘requested X, got Y’, which I assume highlights the issue. 

I’ll be running the same test with lsmash. 

If this is the case, and D2VWitch doesn’t seem to support H264 in the latest versions, does this mean it’s currently impossible to guarantee frame correctness, at least for this BD source?

**Comment by DJATOM** [comment]
Well, for Windows and nvidia GPUs we have DGdecodeNV, but for linux there are not so much you can do. To help ffms be more accurate, you can demux video as raw stream.

**Comment by adworacz** [comment]
Hmmm, good to know about DGDecodeNV, but at this point, I'm sticking with Linux for the time being.

Running the seek-test.py script that @theChaosCoder pointed me to is showing that ffms2 is not frame accurate at all for this source BUT L-SMASH actually is! The test isn't complete yet (~30% complete), but it hasn't found a single frame discrepancy so far, while ffms2 was lighting up like a christmas tree at 1%. Doing a quick test, Av1an seems to produce stable content using L-SMASH as well, so I have high hopes. 

Looks like L-SMASH might be the answer, at least for some of my Bluray content (and as it turns out, MPEG2 DVD content).

**Comment by adworacz** [comment]
I'm going to resolve this issue for now, since it's looking like an issue caused by FFMS2 more than anything.

I'll reopen if I can reproduce issues with L-SMASH, even if it proves to be frame accurate.
